File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        st = []
        for i, c in enumerate(s):
            if c not in "()":
                continue
            if c == ')' and st and st[-1][0] == '(':
                st.pop()
            else:
                st.append((c, i))
        # print(st)
        ans = []
        j = 0
        for i, c in enumerate(s):
            if j < len(st) and i == st[j][1]:
                j += 1
            else:
                ans.append(c)
        return "".join(ans)

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        stack<int> idx;
        for (int i=0; i<s.size(); i++) {
            if (s[i]=='(') idx.push(i);
            else if (s[i]==')') {
                if (idx.empty()) {
                    s.erase(s.begin()+i);
                    i--;
                } else idx.pop();
            } else continue;
        }
        while (!idx.empty()) {
            s.erase(s.begin()+idx.top());
            idx.pop();
        }
        return s;
    }
};